amd_tressfx_vulkan/src/TressFXMeshVulkan.cpp

```diff
@@ -0,0 +1,1147 @@
+    m_HairAsset.m_pTangents = NULL;
+    m_HairAsset.m_pThicknessCoeffs = NULL;
+    m_HairAsset.m_pRestLengths = NULL;
+    m_HairAsset.m_pFollowRootOffset = NULL;
+
+    m_HairAsset.m_NumFollowHairsPerGuideHair = 4;
+    m_HairAsset.m_NumGuideHairStrands = 0;
+    m_HairAsset.m_NumOfVerticesInStrand = 2;
+
+    m_HairSkinMappingBuffer = NULL;
+    m_HairTransformsBuffer = NULL;
+    m_HairSkinMappingView = NULL;
+    m_HairTransformsView = NULL;
+    m_pStrandTexCoordBuffer = NULL;
+    m_pStrandTexCoordView = NULL;
+}
+
+//--------------------------------------------------------------------------------------
+//
+// Destructor
+//
+//--------------------------------------------------------------------------------------
+TressFXMesh::~TressFXMesh(void) { OnDestroy(); }
+
+//--------------------------------------------------------------------------------------
+//
+// OnCreate
+//
+//  Called to create a hair mesh. This function creates the VK resources
+//  and stores the hair data into these resources.
+//
+//--------------------------------------------------------------------------------------
+VkResult TressFXMesh::OnCreate(VkDevice pvkDevice, TressFX_HairBlob *pHairBlob,
+                               TressFX_SceneMesh *sceneMesh, VkImageView pTexture,
+                               uint32_t texture_buffer_memory_index,
+                               VkCommandBuffer upload_cmd_buffer, VkBuffer scratchBuffer,
+                               VkDeviceMemory scratchMemory,
+                               VkDescriptorSetLayout GlobalConstraintsSetLayout,
+                               VkDescriptorSetLayout LocalConstraintsSetLayout,
+                               VkDescriptorSetLayout LenghtWindCollisionSetLayout,
+                               VkDescriptorSetLayout PrepareFollowHairSetLayout,
+                               VkDescriptorSetLayout UpdateFollowHaitSetLayout,
+                               VkDescriptorSetLayout computeTangentSetLayout,
+                               VkDescriptorSetLayout Pass1SetLayout,
+                               VkDescriptorSetLayout ShadowSetLayout)
+{
+    VkResult vr;
+
+    // load the binary file
+    if (!Deserialize(pHairBlob))
+    {
+        return VK_ERROR_INITIALIZATION_FAILED;
+    }
+
+    size_t sizeToUpload = m_HairAsset.m_NumTotalHairVertices * sizeof(float);
+
+    void *uploadBuffer;
+    size_t offsetInUploadBuffer = 0;
+    vkMapMemory(pvkDevice, scratchMemory, 0, sizeToUpload, 0, &uploadBuffer);
+
+    // thickness coeff buffer
+    {
+        VkBufferCreateInfo bd{VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO};
+        bd.usage =
+            VK_BUFFER_USAGE_UNIFORM_TEXEL_BUFFER_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT;
+        bd.size = sizeof(float) * m_HairAsset.m_NumTotalHairVertices;
+
+        AMD_V_RETURN(vkCreateBuffer(pvkDevice, &bd, nullptr, &m_pThicknessCoeffsBuffer));
+        m_pThicknessCoeffsMemory = allocBufferMemory(pvkDevice, m_pThicknessCoeffsBuffer,
```

Comment:
I'd suggest allocating one block of memory for most of those buffers, instead of tying the allocation to the buffer directly. Unless those buffers are in the 60+ MiB range (where it doesn't matter), it's better to have them all in a single allocation.